src/ring_theory/perfection.lean

```diff
@@ -0,0 +1,422 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import algebra.char_p
+import algebra.ring.pi
+import analysis.special_functions.pow
+import field_theory.perfect_closure
+import ring_theory.localization
+import ring_theory.subring
+import ring_theory.valuation.integers
+
+/-!
+# Ring Perfection and Tilt
+
+In this file we define the perfection of a ring of characteristic p, and the tilt of a field
+given a valuation to `ℝ≥0`.
+
+## TODO
+
+Define the valuation on the tilt, and define a characteristic predicate for the tilt.
+
+-/
+
+universes u v w
+
+open_locale nnreal
+
+/-- The perfection of a monoid `M`, defined to be the projective limit of `M`
+using the `p`-th power maps `M → M` indexed by the natural numbers, implemented as
+`{ f : ℕ → M | ∀ n, f (n + 1) ^ p = f n }`. -/
+def submonoid.perfection (M : Type u) [comm_monoid M] (p : ℕ) : submonoid (ℕ → M) :=
+{ carrier := { f | ∀ n, f (n + 1) ^ p = f n },
+  one_mem' := λ n, one_pow _,
+  mul_mem' := λ f g hf hg n, (mul_pow _ _ _).trans $ congr_arg2 _ (hf n) (hg n) }
+
+/-- The perfection of a semiring `R` with characteristic `p`,
+defined to be the projective limit of `R` using the Frobenius maps `R → R`
+indexed by the natural numbers, implemented as `{ f : ℕ → R | ∀ n, f (n + 1) ^ p = f n }`. -/
+def perfection_s (R : Type u) [comm_semiring R] (p : ℕ) [hp : fact p.prime] [char_p R p] :
+  subsemiring (ℕ → R) :=
+{ zero_mem' := λ n, zero_pow $ hp.pos,
+  add_mem' := λ f g hf hg n, (frobenius_add R p _ _).trans $ congr_arg2 _ (hf n) (hg n),
+  .. perfection_m R p }
+  
+/-- The perfection of a ring `R` with characteristic `p`,
+defined to be the projective limit of `R` using the Frobenius maps `R → R`
+indexed by the natural numbers, implemented as `{ f : ℕ → R | ∀ n, f (n + 1) ^ p = f n }`. -/
+def perfection (R : Type u) [comm_ring R] (p : ℕ) [hp : fact p.prime] [char_p R p] :
+  subring (ℕ → R) :=
+{ neg_mem' := λ f hf n, (frobenius_neg R p _).trans $ congr_arg _ (hf n),
+  .. perfection_s R p }
+
+namespace ring_char
+
+instance char_p (R : Type u) [semiring R] : char_p R (ring_char R) :=
+⟨spec R⟩
+
+theorem of_eq {R : Type u} [semiring R] {p : ℕ} (h : ring_char R = p) : char_p R p :=
+h ▸ ring_char.char_p R
+
+theorem eq_iff {R : Type u} [semiring R] {p : ℕ} : ring_char R = p ↔ char_p R p :=
+⟨of_eq, eq.symm ∘ eq R⟩
+
+theorem dvd {R : Type u} [semiring R] {x : ℕ} (hx : (x : R) = 0) : ring_char R ∣ x :=
+(spec R x).1 hx
+
+end ring_char
+
+namespace char_p
+
+instance pi (ι : Type u) [hi : nonempty ι] (R : Type v) [semiring R] (p : ℕ) [char_p R p] :
+  char_p (ι → R) p :=
+⟨λ x, let ⟨i⟩ := hi in iff.symm $ (char_p.cast_eq_zero_iff R p x).symm.trans
+⟨λ h, funext $ λ j, show ring_hom.apply (λ _, R) j (↑x : ι → R) = 0,
+    by rw [ring_hom.map_nat_cast, h],
+  λ h, (ring_hom.apply (λ _, R) i).map_nat_cast x ▸ by rw [h, ring_hom.map_zero]⟩⟩
+
+-- diamonds
```

Comment:
What do you mean by this? `char_p` is a prop, so diamonds don't matter, right?

Reply:
I mean that I need this instance because Lean cannot figure this out because of diamonds.